packages/phpstan-rules/src/Rules/NoModifyAndReturnSelfObjectRule.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Symplify\PHPStanRules\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\Clone_;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\Node\Stmt\Return_;
+use PhpParser\NodeFinder;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\MixedType;
+use Rector\Core\PhpParser\Node\BetterNodeFinder;
+use Symplify\PHPStanRules\NodeFinder\ReturnNodeFinder;
+use Symplify\PHPStanRules\Printer\NodeComparator;
+use Symplify\PHPStanRules\ValueObject\PHPStanAttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Symplify\PHPStanRules\Tests\Rules\NoModifyAndReturnSelfObjectRule\NoModifyAndReturnSelfObjectRuleTest
+ */
+final class NoModifyAndReturnSelfObjectRule extends AbstractSymplifyRule
+{
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = 'Use void instead of modify and return self object';
+
+    /**
+     * @var ReturnNodeFinder
+     */
+    private $returnNodeFinder;
+
+    /**
+     * @var NodeFinder
+     */
+    private $nodeFinder;
+
+    /**
+     * @var NodeComparator
+     */
+    private $nodeComparator;
+
+    public function __construct(ReturnNodeFinder $returnNodeFinder, NodeFinder $nodeFinder, NodeComparator $nodeComparator)
+    {
+        $this->returnNodeFinder = $returnNodeFinder;
+        $this->nodeFinder = $nodeFinder;
+        $this->nodeComparator = $nodeComparator;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        if ($node->params === []) {
+            return [];
+        }
+
+        $returns = $this->returnNodeFinder->findReturnsWithValues($node);
+
+        if ($returns === []) {
+            return [];
+        }
+
+        foreach ($returns as $return) {
+            if (! $return->expr instanceof Expr) {
+                continue;
+            }
+
+            $type = $scope->getType($return->expr);
+            if (! $type instanceof MixedType) {
```

Comment:
@TomasVotruba is there a way to get return type expr type for `Return_::expr` that is an `ObjectType`, seems mostly detect as `MixedType`, eg:

```php
use DateTime;

final class ModifyAndReturnSelfObject
{
    public function run(DateTime $dateTime)
    {
        $dateTime->format('Y-m-d');
        return $dateTime;
    }
}
```

Above, the `$dateTime` detected as `MixedType`

Reply:
I'm looking into this now.

The type was `MixedType`, because the `Scope` was bound to `ClassMethod`, that was used in `getNodeTypes()`.

```php
final class ModifyAndReturnSelfObject
{
    public function run(DateTime $dateTime) // this line
    {
        $dateTime->format('Y-m-d');
        return $dateTime;
    }
}
```

<br>

To use the `Scope` of `Return_`, you have to switch to `Return_` node:

```php
final class ModifyAndReturnSelfObject
{
    public function run(DateTime $dateTime)
    {
        $dateTime->format('Y-m-d');
        return $dateTime; // this line
    }
}
```